core/src/main/java/io/grpc/internal/KeepAliveManager.java

```diff
@@ -0,0 +1,223 @@
+        transport.shutdownNow(Status.UNAVAILABLE);
+      }
+    }
+  };
+  private final Runnable sendPing = new Runnable() {
+    @Override
+    public synchronized void run() {
+      if (state == State.PING_SCHEDULED
+          || (state == State.PING_DELAYED && nextKeepaliveTime <= ticker.read())) {
+        // Send the ping and schedule a shutdown. It fires if we don't receive the ping response
+        // within the timeout.
+        transport.ping(pingCallback, scheduler);
+        state = State.PING_SENT;
+        // Disconnect the transport if we don't get a response within a timeout.
+        shutdownFuture = scheduler.schedule(shutdown, keepAliveTimeoutInNanos,
+            TimeUnit.NANOSECONDS);
+      } else if (state == State.PING_DELAYED) {
+        // We have received some data. Reschedule the ping with the new time.
+        pingFuture = scheduler.schedule(sendPing, nextKeepaliveTime - ticker.read(),
+            TimeUnit.NANOSECONDS);
+        state = State.PING_SCHEDULED;
+      }
+    }
+  };
+  private final KeepAlivePingCallback pingCallback = new KeepAlivePingCallback();
+  private long keepAliveDelayInNanos;
+  private long keepAliveTimeoutInNanos;
+
+  private enum State {
+    /*
+     * Transport has no active rpcs. We don't need to do any keepalives.
+     */
+    IDLE,
+    /*
+     * We have scheduled a ping to be sent in the future. We may decide to delay it if we receive
+     * some data.
+     */
+    PING_SCHEDULED,
+    /*
+     * We need to delay the scheduled keepalive ping.
+     */
+    PING_DELAYED,
+    /*
+     * The ping has been sent out. Waiting for a ping response.
+     */
+    PING_SENT,
+    /*
+     * Transport goes idle after ping has been sent.
+     */
+    IDLE_AND_PING_SENT,
+    /*
+     * The transport has been disconnected. We won't do keepalives any more.
+     */
+    DISCONNECTED,
+  }
+
+  public KeepAliveManager(ManagedClientTransport transport, ScheduledExecutorService scheduler) {
+    this(transport, scheduler, SYSTEM_TICKER, KEEPALIVE_DELAY, KEEPALIVE_TIMEOUT);
+  }
+
+  @VisibleForTesting
+  KeepAliveManager(ManagedClientTransport transport, ScheduledExecutorService scheduler,
+                          Ticker ticker, long keepAliveDelayInNanos, long keepAliveTimeoutInNanos) {
+    this.transport = Preconditions.checkNotNull(transport, "transport");
+    this.scheduler = Preconditions.checkNotNull(scheduler, "scheduler");
+    this.ticker = Preconditions.checkNotNull(ticker, "ticker");
+    this.keepAliveDelayInNanos = keepAliveDelayInNanos;
+    this.keepAliveTimeoutInNanos = keepAliveTimeoutInNanos;
+    nextKeepaliveTime = ticker.read() + keepAliveDelayInNanos;
+  }
+
+  /**
+   * Transport has received some data so that we can delay sending keepalives.
+   */
+  public synchronized void onDataReceived() {
+    nextKeepaliveTime = ticker.read() + keepAliveDelayInNanos;
+    // We do not cancel the ping future here. This avoids constantly scheduling and cancellation in
```

Comment:
My channel idleness implementation (https://github.com/grpc/grpc-java/pull/1986) has the similar situation, and currently I am just cancelling the future. Maybe I should do the same optimization? If I do so, maybe we can share the logic by putting the logic in a small class:
```java
abstract class ResettableTimer {
  abstract Object getLock();

  @GuardeBy("getLock()")
  abstract void timerExpired();

  abstract ScheduledExecutorService getSchdueldExecutor();

  abstract Stopwatch getStopwatch();

  // Stop the timer from expiring
  void suspend();

  // Reset the timer so its expiration time is counted against now
  void reset();
}
```

This would also simplify KeepAliveManager by getting rid of `PING_DELAYED`.